lib/srfi-117.stk

```diff
@@ -0,0 +1,261 @@
+;;;; This program is free software; you can redistribute it and/or modify
+;;;; it under the terms of the GNU General Public License as published by
+;;;; the Free Software Foundation; either version 3 of the License, or
+;;;; (at your option) any later version.
+;;;;
+;;;; This program is distributed in the hope that it will be useful,
+;;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;;;; GNU General Public License for more details.
+;;;;
+;;;; You should have received a copy of the GNU General Public License
+;;;; along with this program; if not, write to the Free Software
+;;;; Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
+;;;; USA.
+;;;;
+;;;; This file is a derivative work from the  implementation of
+;;;; this SRFI by John Cowan, it is copyrighted as:
+;;;;
+;;;;;;  Copyright (c) 2014 John Cowan.
+;;;;;;  Permission is hereby granted, free of charge, to any person
+;;;;;;  obtaining a copy of this software and associated documentation
+;;;;;;  files (the "Software"), to deal in the Software without
+;;;;;;  restriction, including without limitation the rights to use,
+;;;;;;  copy, modify, merge, publish, distribute, sublicense, and/or
+;;;;;;  sell copies of the Software, and to permit persons to whom the
+;;;;;;  Software is furnished to do so, subject to the following
+;;;;;;  conditions:
+;;;;;;
+;;;;;; The above copyright notice and this permission notice shall be
+;;;;;; included in all copies or substantial portions of the Software.
+;;;;;;
+;;;;;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+;;;;;; EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+;;;;;; OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+;;;;;; NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+;;;;;; HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+;;;;;; WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+;;;;;; FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+;;;;;; OTHER DEALINGS IN THE SOFTWARE.
+;;;;
+;;;;           Author: Jeronimo Pellegrini [j_p@aleph0.info]
+;;;;    Creation date: 30-Jun-2020 05:01 (jpellegrini)
+;;;; Last file update: 30-May-2020 06:10 (jpellegrini)
+;;;;
+
+(require "srfi-9")
+
+(define-module SRFI-117
+  (export make-list-queue list-queue list-queue-copy list-queue-unfold list-queue-unfold-right
+          list-queue? list-queue-empty?
+          list-queue-front list-queue-back list-queue-list list-queue-first-last
+          list-queue-add-front! list-queue-add-back! list-queue-remove-front! list-queue-remove-back!
+          list-queue-remove-all! list-queue-set-list!
+          list-queue-append list-queue-append! list-queue-concatenate
+          list-queue-map list-queue-map! list-queue-for-each)
+ 
+;;; This definition is from Chibi's SRFI-1 implementation.
+
+(define (last-pair ls) (if (null? (cdr ls)) ls (last-pair (cdr ls))))
```

Comment:
You can probably delete the definition of `last-pair` since it is is a primitive in STklos (and it is compatible with this definition).

Reply:
OK, I have deleted it and it passes all tests.

Reply:
One thing: it defines map!. If the user does

```
(require "srfi-1")
(requie "srfi-117")
```

Then SRFI-1's `map!` would be shadowed. Is it better to make SRFI-117 require SRFI-1, or to rename `map!` in SRFI-117?